tests/cpp/tree/gpu_hist/test_row_partitioner.cu

```diff
@@ -178,4 +183,34 @@ void TestExternalMemory() {
 }  // anonymous namespace
 
 TEST(RowPartitioner, LeafPartitionExternalMemory) { TestExternalMemory(); }
+
+namespace {
+void TestEmptyNode(std::int32_t n_workers) {
+  collective::TestDistributedGlobal(n_workers, [] {
+    auto ctx = MakeCUDACtx(DistGpuIdx());
+    RowPartitioner partitioner;
+    bst_idx_t n_samples = (collective::GetRank() == 0) ? 0 : 1024;
+    bst_idx_t base_rowid = 0;
+    partitioner.Reset(&ctx, n_samples, base_rowid);
+    std::vector<RegTree::Node> splits(1);
+    partitioner.UpdatePositionBatch(
+        &ctx, {0}, {1}, {2}, splits,
+        [] XGBOOST_DEVICE(bst_idx_t ridx, std::int32_t /*nidx_in_batch*/, RegTree::Node) {
+          return ridx < 3;
+        });
+    ASSERT_EQ(partitioner.GetNumNodes(), 3);
```

Comment:
[Cumulative sum of left and right indices](https://github.com/dmlc/xgboost/blob/master/src/tree/gpu_hist/row_partitioner.cuh#L319), in line 197 above: `1 + 2 = 3`.

Reply:
The `n_nodes_` internal variable is used for sanity checks, it's initialized with 1 with the root node included. The reasoning of choice is that we don't update the partitioner if there's no split for the root node, but the `n_nodes >= 1` is an invariant for all cases since we must have a root node.